batteries/difftext/printdiff.luau

```
@@ -33,22 +37,82 @@ local function visualizeCharDiff(a: string, b: string): (string, string)
+			table.insert(srcVisual, bgRed(op.text))
 		elseif op.key == "ADD" then
-			table.insert(destVisual, richterm.bgGreen(op.text))
+			table.insert(destVisual, bgGreen(op.text))
 		end
 	end
 
 	return table.concat(srcVisual, ""), table.concat(destVisual, "")
 end
 
-local function printDiffByLineDetailed(a: string, b: string): string
-	local diff = myersdiffbyline(a, b)
+local function formatLineSideHeader(
+	operation: types.diffoperationkey,
+	maxNumWidth: number?,
+	lineNumbers: {
+		oldLine: number?,
+		newLine: number?,
+	}?
+): string
+	-- side header structure:
+	-- 	  if line numbers included:
+	-- 		ALWAYS has length maxNumWidth*2 + 4
+	-- 		the padded structure is essentialy:
+	--		{operation} {oldLine# padded to len of maxNumWidth} {newLine# padded to len = maxNumWidth}|
+	-- 	  if line numbers excluded:
+	-- 		"+ " for additions, "- " for deletions, "" for equality
+
+	local function padding(width: number)
+		return (" "):rep(width)
+	end
+
+	local oldStr = tostring(if lineNumbers then lineNumbers.oldLine else "")
+	local newStr = tostring(if lineNumbers then lineNumbers.newLine else "")
+	maxNumWidth = maxNumWidth or 1
+	if operation == "EQUAL" then
+		if not lineNumbers then
+			return ""
+		end
+		local leftPad = maxNumWidth - #oldStr
+		local rightPad = maxNumWidth - #newStr
+		return string.format("  %s%s %s%s| ", padding(leftPad), oldStr, padding(rightPad), newStr)
+	elseif operation == "ADD" then
+		if not lineNumbers then
+			return "+ "
+		end
+		local leftPad = maxNumWidth -- space for missing oldLine
+		local rightPad = maxNumWidth - #newStr
+		return string.format("+ %s %s%s| ", padding(leftPad), padding(rightPad), newStr)
+	elseif operation == "DELETE" then
+		if not lineNumbers then
+			return "- "
+		end
+		local leftPad = maxNumWidth - #oldStr
+		local rightPad = maxNumWidth -- space for missing newLine
+		return string.format("- %s%s %s| ", padding(leftPad), oldStr, padding(rightPad))
+	end
+
+	return "" -- unreachable
```

it might be easier to debug in the future if we `error()` here instead of failing silently